xapian-core/include/xapian/mset.h

```diff
@@ -83,25 +82,24 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
      * It updates weights of all the elements in items as well as updating the max_attained
      */
     template <typename Iterator>
-    void set_new_weights(Iterator first, Iterator last) {
-	int i = 0;
+    void replace_weights(Iterator first, Iterator last) {
 	if (std::distance(first, last) != size())
 	{
-	    printf("Number of weights assigned don't match the number of items.Aborting");
+	    throw std::invalid_argument("Number of weights assigned don't match the number of items.Aborting");
 	    return;
 	}
-	for (Iterator it = first;it != last;++it)
+	int i = 0;
+	for (Iterator it = first;it != last;++it,++i)
 	{
-	    set_new_weights_helper(*it, i);
-	    ++i;
+	    set_item_weight(i, (*it).get_score());
```

Comment:
This isn't a great idea, because then the only iterators that make sense are ones that have `.get_score()`. You should be able to create a custom iterator inside `ranker/ranker.cc` which dereferences to `double` (in other words, which has a `double operator*()` method). Then the public API for changing weights can be used by anyone with a `vector<double>`, for instance, which is more what I'd expect.